proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,397 @@
+   ::
+     <decl> ::= $let <integer-literal> <<existing syntax>> = <<existing syntax>>
+   In both case the ``$`` must not be followed by whitespace, both to avoid conflicts with other syntax and to be consistent with splices.
+   The means bind identifiers in stage *n* instead of stage 0 as per normal.
+   Module exports however are restricted to stage 0.
+
+#. The current "stage restriction" on splices using items from module is abolished.
+   Any stage n + 1 binding in a stage n splice is fair game.
+
+#. Relax ``-XTemplateHaskellQuotes`` to instead allow Template Haskell constructs, but restrict their usage so all syntax is in stages >= 0.
+
+#. Introduce ``-XTemplateStagePersistence``.
+   Which is implied by ``-XTemplateHaskellQuotes`` (and thus plain ``-XTemplateHaskell``) for backwards compatibility.
+   It allows the current behavior where we blur the distinction between stages.
+   In particular, with this enabled:
+
+   - Stage 0 identifiers bound in another module can be used in stage 1 (quotes) and stage -1 (splices).
+   - Stage 0 identifiers bound anywhere can be used in stage 1, and are automatically.
+   - Typed template haskell is allowed.
+   - The ``Lift`` type class and all its associated definitions are made available.
+
+   These are all already always permitted today.
+   But with ``-XNoTemplateStagePersistence``, overriding the default, all of those are *disabled*.
+
+#. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.
+   If/when GHC becomes multi-target, by default stages >= 0 take GHC's target platform / the packages host platform (where compiled code runs), while stages < 0 take GHC's host platform / the packages build platform (where GHC runs).
+   But, the emitted platform can still be specified per-stage like the other flags.
+   This is needed when building TH functions to be used from cross compiled code.
+
+#. When importing modules/packages, after applying the import offset ensure that the platforms match.
+   Note that while each module only has bindings in its own stage 0, those bindings can contain quotes from stages greater than 0.
+   All such quoted platforms need to match.
+
+#. Just as GHC defines ``*_HOST_OS`` and similar CPP identifiers today, it would define ``*_BUILD_*`` ones if you have any stage -1 package imports, and ``*_BUILD_*`` if you have any stage 1 package imports.
+
+Cabal
+~~~~~~~~~~~~
+
+#. Extend the ``build-depends`` syntax with a stage integer offset parameter.
+   N.B ``build-tool-depends`` can be thought of as a stage -1 executable dependencies list.
+   `<https://github.com/haskell/cabal/issues/5411>`_ asks for a ``run-tool-depends`` which would be nothing but a stage 0 executable depends.
+   ``setup-depends`` can also be thought of as a stage -1 executable dependencies list.
+
+#. Likewise extend ``other-modules`` with a stage integer offset parameter, to support intra-package ``$import``.
```

Comment:
You can have a -1 `other-module` which is just used for a splice, rather than exported. But yeah I am not wild about this feature and wouldn't mind dropping it.